R Analysis
================
Jennifer Chang
1/8/2021

## R Markdown

This is a placeholder to store any analysis in R (edgeR, DESeq2). Also a
way to look at the underlying assumptions (RNASeq distribution functions
and how it behaves in an analysis pipeline. Is the distribution
realistic?)

## Install any required libraries

``` r
# == Install any CRAN packages
# cran_pkgs <- c("QuasiSeq")
# install.packages(cran_pkgs)

# == Install any Bioconductor packages
# bioc_pkgs <- c("edgeR", "baySeq", "DESeq2", "ballgown", "NOISeq", "limma")
# BiocManager::install(bioc_pkgs)

# ====================   Year of development
library(tidyverse)
library(magrittr)
library(edgeR)         # 2010
library(baySeq)        # 2010
library(QuasiSeq)      # 2012
library(DESeq2)        # 2014    # <= actually a bit strange that DESeq2 & edgeR tend to be equivalent, despite being 4yrs later
library(ballgown)      # 2014
library(NOISeq)        # 2015
library(limma)         # 2015
```

<!--  ## For now comment out exploration portion.
## Distribution

RNASeq is assumed to be [negatively binomially distributed](https://en.wikipedia.org/wiki/Negative_binomial_distribution) according to the edgeR and DESeq2 papers. 

$$
Y_{gi} \sim NB(M_{i}p{gj}, \phi_{g})
$$

What does a negative binomial distribution look like?


```r
y  <- rnbinom( n = 1000,       # Randomly generate 1000 entries (number of replicates?)
               size = 5,       # at least 5 successes (expressed/not?)
               prob = 0.5      # probability of success in each case...maybe length of genome may affect this
               )

df <- data.frame(x = c(1:1000),
                 y = y)
                  
ggplot(data = df, aes(x=y)) + 
  geom_histogram(binwidth = 1) +
  theme_bw() +
  labs(title="Negative Binomial Distribution (n=1000, size=5, prob=0.5)") +
  geom_vline(xintercept=5, color="blue")
```

![](imgs/2_DE_analysisnegbinomial-1.png)<!-- -->

Notice how the peak of the distribution is around 5 (`size=5`). Play
around with the `rnbinom` parameters, how does that change the shape of
the distribution? What implications does this have for RNASeq to
determine if a gene is differentially expressed across two treatments?
–&gt;

# Load in RNAseq counts

Load in the RNAseq counts generated by GSNAP.

``` r
data_maize <- read_delim("maize_genecounts.txt", delim = "\t")
#> 
#> ── Column specification ────────────────────────────────────────────────────────
#> cols(
#>   .default = col_double(),
#>   Geneid = col_character()
#> )
#> ℹ Use `spec()` for the full column specifications.
data_bee <- read_delim("bee_genecounts.txt", delim="\t")
#> 
#> ── Column specification ────────────────────────────────────────────────────────
#> cols(
#>   .default = col_double(),
#>   Geneid = col_character()
#> )
#> ℹ Use `spec()` for the full column specifications.
```

# Diagnostic plots

## Bee

``` r
melt_bee <- data_bee %>%
  pivot_longer(!Geneid, names_to = "replicate", values_to = "count") %>%
  mutate(
    group=substr(replicate, 1, 3)                 # <= replace this with metadata information later
  )

melt_bee %>% 
  ggplot(., aes(x=replicate, y=count, group=Geneid)) + 
  geom_line(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90, vjust=0.5)
  ) +
  labs(title="GSNAP RNAseq Gene Profiles")
```

![](imgs/2_DE_analysisbee_linechart-1.png)<!-- -->

Distribution, violin plot

``` r
melt_bee %>% 
  ggplot(., aes(x=replicate,y=log(count))) +       # <= log transformed as a rough normalization... we will use DESeq2's normalization eventually
  geom_violin(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90)
  ) +
  labs(title="GSNAP RNAseq Gene Profiles") +
  facet_wrap(~group, scales="free", drop=T)
#> Warning: Removed 216873 rows containing non-finite values (stat_ydensity).
```

![](imgs/2_DE_analysisviolin-1.png)<!-- -->

Check if any of the replicates seem obviously different from the others.
Mostly seems fine, `1-E07-F3-S59 distribution` is a little low compared
to others, `1-B03-A16_S21` seems a little high compared to others.

## Maize

``` r
melt_maize <- data_maize %>%
  pivot_longer(!Geneid, names_to = "replicate", values_to = "count")

melt_maize %>% 
  ggplot(., aes(x=replicate, y=count, group=Geneid)) + 
  geom_line(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90, vjust=0.5)
  ) +
  labs(title="GSNAP RNAseq Gene Profiles")
```

![](imgs/2_DE_analysismaize_linechart-1.png)<!-- -->

``` r
melt_maize %>% 
  ggplot(., aes(x=replicate,y=log(count))) +       # <= log transformed as a rough normalization... we will use DESeq2's normalization eventually
  geom_violin(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90)
  ) +
  labs(title="GSNAP RNAseq Gene Profiles")
#> Warning: Removed 473515 rows containing non-finite values (stat_ydensity).
```

![](imgs/2_DE_analysismaize_violin-1.png)<!-- -->

``` r
#  facet_wrap(~group, scales="free", drop=T)
```

Check if any of the replicates seem obviously different from the others.
Mostly seems okay, `SRR1573509` and `SRR1573523` is a little low
compared to others, `SRR1573513` seems a little high compared to others.
