---
title: "DESeq2 Analysis of Bee and Maize RNAseq data"
author: "Jennifer and Mou"
date: "Updated 4/7/2021"
output:
  github_document
editor_options:
  chunck_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "imgs/DESeq2_analysis"
)
```

## RNA analysis in DESeq2 on bee and maize data.

## Install any required libraries
```{r, warning=FALSE, message=FALSE}
# == Install any CRAN packages
# cran_pkgs <- c("QuasiSeq")
# install.packages(cran_pkgs)

# == Install any Bioconductor packages
# bioc_pkgs <- c("edgeR", "baySeq", "DESeq2", "ballgown", "NOISeq", "limma")
# BiocManager::install(bioc_pkgs)
library(tidyverse)
library(magrittr)
library(DESeq2) 
library(wesanderson)
```



# Load in RNAseq counts
Load in the RNAseq counts generated by GSNAP.
```{r}
data_maize <- read_delim("~/Desktop/maize/maize.genecounts.out.txt", delim = "\t")
data_bee <- read_delim("~/Desktop/bee/bee.genecounts.out.txt", delim="\t")
```

# Diagnostic plots
These are quick plots to look at the data and notice if any of the counts seem unusually high or low, or multimodal. It can indicate a problem with a sample, but does not necessarily mean there is a problem with a sample.

## Bee
```{r bee_linechart, fig.width=8, fig.height=5}
melt_bee <- data_bee %>%
  pivot_longer(!Geneid, names_to = "replicate", values_to = "count")  %>%
  mutate(
    group=substr(replicate, 1, 3)                 # <= replace this with metadata information later (A1-A19, B1-B*)
  )

melt_bee %>% 
  ggplot(., aes(x=replicate, y=count, group=Geneid)) + 
  geom_line(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90, vjust=0.5)
  ) +
  labs(title="GSNAP RNAseq Bee Count Profiles") #try boxplot with group?
ggsave("Bee_GSNAP_RNAseq_CountProfiles_Linechart.png")
```

```{r bee_violin, fig.width=11, fig.height=8}
melt_bee %>% 
  ggplot(., aes(x=replicate,y=log(count))) +       # <= log transformed as a rough normalization... we will use DESeq2's normalization eventually
  geom_violin(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90)
  ) +
  labs(title="GSNAP distribution of bee gene counts by sample") +
  facet_wrap(~group, scales="free", drop=T)
ggsave("Bee_GSNAP_RNAseq_LogCountProfiles_ViolinPlot.png")
```

Check if any of the replicates seem obviously different from the others. Mostly seems fine, `1-E07-F5-S61` (exposed) is a little low compared to others, missing `1-B03-A16_S21` (control).

## Maize
```{r maize_linechart, fig.width=8, fig.height=5,eval=FALSE}
melt_maize <- data_maize %>%
  pivot_longer(!Geneid, names_to = "replicate", values_to = "count")

melt_maize %>% 
  ggplot(., aes(x=replicate, y=count, group=Geneid)) + 
  geom_line(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90, vjust=0.5)
  ) +
  labs(title="GSNAP RNAseq Maize Count Profiles")
ggsave("Maize_GSNAP_RNAseq_CountProfiles_Linechart.png")
```

```{r maize_violin, fig.width=8, fig.height=4, eval=FALSE}
melt_maize %>% 
  ggplot(., aes(x=replicate,y=log(count))) +       # <= log transformed as a rough normalization... we will use DESeq2's normalization eventually
  geom_violin(alpha=0.5) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=90)
  ) +
  labs(title="GSNAP distribution of maize gene counts by sample")
#  facet_wrap(~group, scales="free", drop=T)
ggsave("Maize_GSNAP_RNAseq_LogCountProfiles_ViolinPlot.png")
```

Check if any of the replicates seem obviously different from the others. Mostly seems okay, `SRR1573523` and `SRR1573525` is a little low compared to others, `SRR1573513` seems a little high compared to others.

# DESEQ2
Most of these commands Jennifer got from Siva.

##  Bee and Maize
Prepare data for DESeq2
```{r}
#library(DESeq2)

# bee
de_input = as.matrix(data_bee[-1])
row.names(de_input) = data_bee$Geneid
de_input[1:5,1:6]

metadata <- readr::read_delim("~/Desktop/bee/bumblebee_meta.csv", delim=",")
row.names(metadata) = metadata$Well_ID

meta_df <- data.frame( Sample = colnames(de_input)) %>%
  separate(Sample, 
           c(NA, "group", NA, NA), #NA doesn't create column
           sep="-", 
           remove=FALSE) %>%
  mutate(
    Condition = metadata[group,]$Trt,
    Condition = factor(Condition, levels=c("ctrl", "exposed"))) # factor help save memory, string saved as numbers. Deseq2 will take first factor as "wildtype"); can tell ggplot what order to list groups in legend
# ignore missing pieces
# metadata["A01",]$Trt <= gives you "ctrl"
# metadata["B07",]$Trt <= gives you "exposed"

# Played around with mutate function, seeing what happens when I change metadata[group,]$Trt to metadata[Well_ID,]$Nest. 
# Then decided to merge test2 and metadata dataframes to get Trt and Nest columns in test2 dataframe 
test <- data.frame( Sample = colnames(de_input)) %>%
  separate(Sample, 
           c(NA, NA, "SampleName"), #NA doesn't create column
           sep="-", 
           remove=FALSE) 
test2 <- test %>% separate(SampleName,
                  c("Sample_Name", NA, NA, NA, NA),
                  sep = "_",
                  remove = FALSE)
test3 <- merge(metadata, test2, by = "Sample_Name")
test4 <- test3[-c(6)] #drop SampleName column
test4$Trt <- factor(test4$Trt, levels = c("ctrl", "exposed"))

# maize
de_inputm = as.matrix(data_maize[-1])
row.names(de_inputm) = data_maize$Geneid
de_inputm[1:5,1:6]

# Modify metadata for maize
# download metadata as tsv from https://www.ebi.ac.uk/ena/browser/view/PRJNA260793
# modify metadata so that the sample_tissue entries are converted to the letters only (B-3 becomes B, L-4 becomes L, wtL-1 becomes wtL, L_L1.3 becomes L_L)
metadatam <- readr::read_delim("~/Desktop/maize/maize_metadata.csv", delim=",") 
#row.names(metadatam) = metadatam$run_accession 
names(metadatam)[names(metadatam) == "run_accession"] <- "replicate"  #rename column to match with melt_maize
meta_maize <- merge(metadatam, melt_maize, by = "replicate")
```

Start DESeq2, modified from Masonbrink's notes.

## Bee PCA by treatment group
```{r pca_bee, fig.width=5, fig.height=3}
#BiocManager::install("apeglm")
#library(apeglm)
# bee
dds <- DESeqDataSetFromMatrix(
  de_input,                # Gene counts
  colData = meta_df,       # Metadata
  design = ~Condition      # experimental design
) #set up specialized object
dds <- DESeq(dds) #actually running deseq
res <- results(dds)
table(res$padj<0.05) #looking at which ones are sig. diff. from control at 95% CI
#TRUE = 217 genes are diff. expressed
#FALSE = 8292
res <- res[order(res$padj), ] #come back to this later (take it apart). Order orders it in ascending order
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
write.csv(resdata, file="Bee_AllExposedvsAllControlGene_DeSeq2.csv",quote = FALSE,row.names = F)
resultsNames(dds) #"Intercept"           "Trt_exposed_vs_ctrl"
rld <- rlogTransformation(dds) #vst() much faster transformation ... check it out
summary(rld)
# "DESeqTransform object of length 12925 with 24 metadata columns"
(p <- plotPCA(rld, intgroup="Condition") +
    theme_bw())
ggsave("Bee_AllGenes_PCA_ExposedvsControl.png")

# log fold change shrinkage for visualization and ranking
resultsNames(dds) #[1] "Intercept"                 "Condition_exposed_vs_ctrl"
res2 <- lfcShrink(dds, coef ="Condition_exposed_vs_ctrl", type="apeglm") 
table(res2$padj<0.05)
#TRUE = 217 genes are diff. expressed
#FALSE = 8292
# we get the same # as when we didn't transform data, so this should be ok.
res2 <- res2[order(res$padj), ] #come back to this later (take it apart). Order orders it in ascending order
res2data <- merge(as.data.frame(res2), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(res2data)[1] <- "Gene"
write.csv(res2data, file="Bee_AllExposedvsAllControlGene_DeSeq2_apeglmShrinkageEstimator.csv",quote = FALSE,row.names = F)
# this csv is the exact same info as "Bee_AllExposedvsAllControlGene_DeSeq2.csv"

# filter out only genes with pvalue <=0.05. How to filter them out from dds?
lowresdata <- filter(resdata, pvalue<=0.05)

# with test4 file - it works! Got same results as dds above
dds <- DESeqDataSetFromMatrix(
  de_input,                # Gene counts
  colData = test4,       # Metadata
  design = ~Trt      # experimental design
) #set up specialized object
dds <- DESeq(dds) #actually running deseq
res <- results(dds)
table(res$padj<0.05) #looking at which ones are sig. diff. from control at 95% CI
#TRUE = 217 genes are diff. expressed
#FALSE = 8292
res <- res[order(res$padj), ] #come back to this later (take it apart). Order orders it in ascending order
resdata <- merge(as.data.frame(res), as.data.frame(counts(dds, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdata)[1] <- "Gene"
write.csv(resdata, file="Bee_AllExposedvsAllControlGene_DESeq2_2.csv",quote = FALSE,row.names = F)
rld <- rlogTransformation(dds) #vst() much faster transformation ... check it out
summary(rld)
# "DESeqTransform object of length 12925 with 24 metadata columns"
(plotPCA(rld, intgroup="Trt") +
    theme_bw())
a <- plotPCA(rld, intgroup="Trt", returnData=TRUE)
ggplot(a, aes(PC1, PC2, color=group)) +
  geom_point(size=3) +
  coord_fixed()
```

## Bee PCA by nest
```{r pca_bee_nest, fig.width=5, fig.height=3}
# experimental design by nest
ddsnest <- DESeqDataSetFromMatrix(
  de_input,                # Gene counts
  colData = test4,       # Metadata
  design = ~Nest      # experimental design
) #set up specialized object
ddsnest <- DESeq(ddsnest) #actually running deseq
resnest <- results(ddsnest)
table(resnest$padj<0.05) #looking at which ones are sig. diff. from control at 95% CI
#TRUE = 256 genes are diff. expressed
#FALSE = 9080
resnest <- resnest[order(resnest$padj), ] #come back to this later (take it apart). Order orders it in ascending order
resdatanest <- merge(as.data.frame(resnest), as.data.frame(counts(ddsnest, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdatanest)[1] <- "Gene"
write.csv(resdata, file="Bee_Nest_DESeq2.csv",quote = FALSE,row.names = F)

rldnest <- rlogTransformation(ddsnest) #vst() much faster transformation ... check it out
summary(rldnest)
# "DESeqTransform object of length 12925 with 24 metadata columns"
# PCA by nest
b <- plotPCA(rldnest, intgroup="Nest", returnData=TRUE)
percentVar <- round(100 * attr(b, "percentVar")) #To add % variance for PC1 and PC2 @ plot axes
ggplot(b, aes(PC1, PC2, color=as.factor(group))) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  scale_color_brewer(palette = "Dark2") +
  labs(color = "Nest")
ggsave("Bee_AllGenes_PCA_Nest.png")
```

## Maize PCA
```{r pca_maize, fig.width=5, fig.height=3}
# maize
ddsm <- DESeqDataSetFromMatrix(
  de_inputm,                # Gene counts
  colData = metadatam,       # Metadata
  design = ~Tissue      # experimental design
) #set up specialized object
ddsm <- DESeq(ddsm) #actually running deseq
resm <- results(ddsm)
table(resm$padj<0.05) #looking at which ones are sig. diff. from control at 95% CI
#FALSE  20274 
#TRUE 3189  
resm <- resm[order(resm$padj), ] #come back to this later (take it apart). Order orders it in ascending order
resdatam <- merge(as.data.frame(resm), as.data.frame(counts(ddsm, normalized=TRUE)), by="row.names", sort=FALSE)
names(resdatam)[1] <- "Gene"
write.csv(resdatam, file="Maize_Tissue_DeSeq2.csv",quote = FALSE,row.names = F)

rldm <- rlogTransformation(ddsm) #vst() much faster transformation ... check it out
summary(rldm)
# "DESeqTransform object of length 44703 with 47 metadata columns"
(pm <- plotPCA(rldm, intgroup="Tissue") +
    theme_bw() +
    labs(x = "PC1", y="PC2")
) #without PC1, PC2 % variance axes labels

(pm2 <- plotPCA(rldm, intgroup="Tissue") +
    theme_bw()) #with PC1, PC2 variance % axes labels
ggsave("Maize_AllGenes_PCA_Tissue.png")
```

Other things to try:
ANOVA stats to look at treatment, nest effects on variation (only genes with large fold-change, p<0.05)


# Other diagnostic plots
## Bee differentially expressed genes line plot
```{r bee_deline, fig.width=8, fig.height=4}
# bee
ordered_trt = c(meta_df %>% subset(Condition=="ctrl") %>% {.$Sample},
                meta_df %>% subset(Condition=="exposed") %>% {.$Sample})

# All genes (including non-significant ones)
DEgenes <- resdata %>%
  subset(padj < 0.05) %>%
  select(-c(2:7)) %>%
  pivot_longer(col=-1, names_to="treatment", values_to="expression") %>% 
  separate(treatment, 
           c(NA, "group", NA, NA), 
           sep="-", 
           remove=FALSE) %>%
  mutate(
    Condition = metadata[group,]$Trt,
    Condition = factor(Condition, levels=c("ctrl", "exposed")),
    treatment = factor(treatment, levels=ordered_trt)
  )

# Genes with p < 0.05
DEgeneslow <- lowresdata %>%
  subset(padj < 0.05) %>%
  select(-c(2:7)) %>%
  pivot_longer(col=-1, names_to="treatment", values_to="expression") %>% 
  separate(treatment, 
           c(NA, "group", NA, NA), 
           sep="-", 
           remove=FALSE) %>%
  mutate(
    Condition = metadata[group,]$Trt,
    Condition = factor(Condition, levels=c("ctrl", "exposed")),
    treatment = factor(treatment, levels=ordered_trt)
  )

# Very rough draft, fix this later...
DEgeneslow %>% ggplot(., aes(x=treatment, y=expression, group=Gene)) +
  geom_line(alpha=0.5) +
  geom_point(aes(color=Condition), size=0.5) +
  theme_bw() + 
  theme(
    axis.text.x = element_text(angle=90)
  )# +
#  facet_wrap(~Condition, scales="free_x", drop=T)
```
## Bee volcano plot of differentially expressed genes
```{r bee_volplot, fig.width=10, fig.height=4, warning=FALSE}
# Tutorial: https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html

# basic volcano plot, pvalue converted into -log10(p-value) to flip plot
(volbee<- resdata %>% ggplot(., aes(x=log2FoldChange, y=-log10(pvalue))) +
  geom_point(size=0.5) +
  theme_bw())
 
# Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold
(volbee2 <- volbee + geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
  geom_hline(yintercept = -log10(0.05), col='red'))

# Color the genes that are up or down-regulated
resdata$diffexpressed <-  "NO"
# if log2FoldChange > 0.6 and pvalue < 0.05, set as "UP"
resdata$diffexpressed[resdata$log2FoldChange > 0.6 & resdata$pvalue < 0.05] <- "UP"
# if log2FoldChange < -0.6 and pvalue < 0.05, set as "DOWN"
resdata$diffexpressed[resdata$log2FoldChange < -0.6 & resdata$pvalue < 0.05] <- "DOWN"
# Replot and color the points with "diffexpressed"
(volbee <- ggplot(resdata, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed)) + geom_point() + theme_bw())
# Add vertical lines like before
(volbee2 <- volbee + geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
  geom_hline(yintercept = -log10(0.05), col='red'))

# Add names of genes beside the points
resdata$delabel <- NA
resdata$delabel[resdata$diffexpressed != "NO"] <- resdata$Gene[resdata$diffexpressed != "NO"]
levels(resdata$diffexpressed) <- list("Down-regulated" = "DOWN",
                                      "Up-regulated" = "UP",
                                      "Not differentially regulated" = "NO")
levels(resdata$diffexpressed)
(ggplot(resdata, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed, label=delabel)) +
    geom_point() +
    geom_text(size=2) +
    geom_hline(yintercept = -log10(0.05), col='red') +
    geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
    theme_bw() +
    labs(color = "Differentially and \nNon-differentially \nExpressed Genes") +
    scale_color_discrete(labels = c("Down-regulated", "Not regulated", "Up-regulated"))
    )

ggsave("Bee_AllGenes_VolcanoPlot.png")

```

## Maize differentially expressed genes line plot
```{r maize_deline, fig.width=8, fig.height=4}
# maize
ordered_trtm = c(metadatam %>% subset(Tissue=="B") %>% {.$replicate},
                metadatam %>% subset(Tissue=="L") %>% {.$replicate},
                metadatam %>% subset(Tissue=="S") %>% {.$replicate},
                metadatam %>% subset(Tissue=="wtL") %>% {.$replicate},
                metadatam %>% subset(Tissue=="B_L") %>% {.$replicate},
                metadatam %>% subset(Tissue=="L_L") %>% {.$replicate},
                metadatam %>% subset(Tissue=="S_L") %>% {.$replicate}
               ) # subset samples from each tissue
DEgenesm <- resdatam %>%
  subset(padj < 0.05) %>%
  select(-c(2:7)) %>%
  pivot_longer(col=-1, names_to="replicate", values_to="expression")
datam <-  merge(metadatam, DEgenesm, by = "replicate")
datam %>% ggplot(., aes(x=Tissue, y=expression, group=Gene)) +
  geom_line(alpha=0.5) +
  geom_point(aes(color=Tissue), size=0.5) +
  theme_bw() + 
  theme(
    axis.text.x = element_text(angle=90)
  )# +
#  facet_wrap(~Condition, scales="free_x", drop=T)
```

## Maize volcano plot of differentially expressed genes
```{r maize_volplot, fig.width=10, fig.height=4, warning=FALSE}
# basic volcano plot, pvalue converted into -log10(p-value) to flip plot
(volm<- resdatam %>% ggplot(., aes(x=log2FoldChange, y=-log10(pvalue))) +
  geom_point(size=0.5) +
  theme_bw())
 
# Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold
(volm2 <- volm + geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
  geom_hline(yintercept = -log10(0.05), col='red'))

# Color the genes that are up or down-regulated
resdatma$diffexpressed <-  "NO"
# if log2FoldChange > 0.6 and pvalue < 0.05, set as "UP"
resdatam$diffexpressed[resdatam$log2FoldChange > 0.6 & resdatam$pvalue < 0.05] <- "UP"
# if log2FoldChange < -0.6 and pvalue < 0.05, set as "DOWN"
resdatam$diffexpressed[resdatam$log2FoldChange < -0.6 & resdatam$pvalue < 0.05] <- "DOWN"
# Replot and color the points with "diffexpressed"
(volm <- ggplot(resdatam, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed)) + geom_point() + theme_bw())
# Add vertical lines like before
(volm2 <- volm + geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
  geom_hline(yintercept = -log10(0.05), col='red'))

# Add names of genes beside the points
resdatam$delabel <- NA
resdatam$delabel[resdatam$diffexpressed != "NO"] <- resdatam$Gene[resdatam$diffexpressed != "NO"] # This didn't seem to work... Will manually call labels in legend
(ggplot(resdatam, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed, label=delabel)) +
    geom_point() +
    geom_text(size=2) +
    geom_hline(yintercept = -log10(0.05), col='red') +
    geom_vline(xintercept=c(-0.6,0.6), col = 'red') +
    theme_bw() +
    labs(color = "Differentially and \nNon-differentially \nExpressed Genes") +
    scale_color_discrete(labels = c("Down-regulated", "Not regulated", "Up-regulated"))
    )

ggsave("Maize_AllGenes_VolcanoPlot.png")

```

Next: plot the top 20 differentially expressed genes (lowest padj values)... go back and check on normalization and metadata...

## Bee top 20 differentially expressed genes line plot
```{r bee_top20, fig.width=8, fig.height=4}
# bee
top20 = resdata[c(1:20),] %>%
  subset(padj < 0.05) %>%
  select(-c(2:7)) %>%
  pivot_longer(col=-1, names_to="treatment", values_to="expression") %>% 
  separate(treatment, 
           c(NA, "group", NA, NA), 
           sep="-", 
           remove=FALSE) %>%
  mutate(
    Condition = metadata[group,]$Trt,
    Condition = factor(Condition, levels=c("ctrl", "exposed")),
    treatment = factor(treatment, levels=ordered_trt)
  )

top20 %>% ggplot(., aes(x = as.factor(Gene), y = expression, fill=Condition))+
  geom_boxplot() + 
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=45, hjust = 1)
  ) +
  labs(x="Gene")
```

## Maize top 20 differentially expressed genes line plot
```{r maize_top20, fig.width=8, fig.height=4}
# maize
top20m = resdatam[c(1:20),] %>%
  subset(padj < 0.05) %>%
  select(-c(2:7)) %>%
  pivot_longer(col=-1, names_to="replicate", values_to="expression")
top20m2 <-  merge(metadatam, top20m, by = "replicate")
top20m2 %>% ggplot(., aes(x=as.factor(Gene), y=expression, fill=Tissue)) +
  geom_boxplot() + 
  theme_bw() +
  theme(
    axis.text.x = element_text(angle=45, hjust = 1)
  ) +
  labs(x="Gene")
```

Now plot heatmap of count matrix

## Bee heatmap of count matrix
```{r bee_heatmap, fig.width=11, fig.height=16, eval=FALSE}
# bee
library("pheatmap")
beeheatmap <- pheatmap(assay(rld)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
ggsave("Bee_HeatmapOfCountMatrix.png", plot=beeheatmap, width=13, height=16)
# How to change colors?
```

## Maize heatmap of count matrix
```{r maize_heatmap, fig.width=11, fig.height=16, eval=FALSE}
# bee
library("pheatmap")
maizeheatmap <- pheatmap(assay(rldm)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
ggsave("Maize_HeatmapOfCountMatrix.png", plot=maizeheatmap, width=13, height=16)
# How to change colors?
```

Heatmap of sample-to-sample distances

## Bee
```{r bee_heatmap2, fig.width=11, fig.height=16, eval=FALSE}
# bee
library("pheatmap")
vsd <- vst(dds, blind = FALSE) #shorter running times if use blind=FALSE
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Condition, vsd$group)
colnames(sampleDistMatrix) <- NULL
beeheatmap2 <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists)
ggsave("Bee_HeatmapOfSampleToSampleDistances.png", plot=beeheatmap2)
```

## Maize
```{r maize_heatmap2, fig.width=11, fig.height=16, eval=FALSE}
# bee
library("pheatmap")
vsdm <- vst(ddsm, blind = FALSE) #shorter running times if use blind=FALSE
sampleDistsm <- dist(t(assay(vsdm)))
sampleDistMatrixm <- as.matrix(sampleDistsm)
rownames(sampleDistMatrixm) <- paste(vsdm$Condition, vsdm$group)
colnames(sampleDistMatrixm) <- NULL
maizeheatmap2 <- pheatmap(sampleDistMatrixm,
         clustering_distance_rows = sampleDistsm,
         clustering_distance_cols = sampleDistsm)
ggsave("Maize_HeatmapOfSampleToSampleDistances.png", plot=maizeheatmap2)
```