# Notes, Ideas and Thoughts

## To do
1. Bee
* Toss out the one sample and run deseq2, see how that looks in R script QC section
* Look through gff file for annotation of gene locus names: https://github.com/ISUgenomics/2021_workshop_transcriptomics/blob/main/Notebook_Masonbrink/02_Bumblebee/07_FeatureTable.md

2. Maize
* Run deseq2, look at QC and deseq2 results

3. PCIT articles:
  * https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000382#s4
  * https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0128350

<details><summary>blah</summary>
dfdfd
</details>

* Run PCA of only diff expressed genes to get better clustering (if only a few genes diff. expressed, would not see clear clusters - get the scatter that Rick saw)
